File: src/main/java/no/nav/tjenestepensjon/simulering/soap/SoapClient.java
package no.nav.tjenestepensjon.simulering.soap;

import static no.nav.tjenestepensjon.simulering.mapper.SoapMapper.mapSimulerTjenestepensjonRequest;
import static no.nav.tjenestepensjon.simulering.mapper.SoapMapper.mapSimulerTjenestepensjonResponse;
import static no.nav.tjenestepensjon.simulering.mapper.SoapMapper.mapStillingsprosentRequest;
import static no.nav.tjenestepensjon.simulering.mapper.SoapMapper.mapStillingsprosentResponse;

import java.util.List;

import org.slf4j.Logger;
import org.slf4j.LoggerFactory;
import org.springframework.stereotype.Component;
import org.springframework.ws.client.core.WebServiceTemplate;
import org.springframework.ws.client.core.support.WebServiceGatewaySupport;

import no.nav.ekstern.pensjon.tjenester.tjenestepensjonsimulering.v1.HentStillingsprosentListeResponse;
import no.nav.ekstern.pensjon.tjenester.tjenestepensjonsimulering.v1.SimulerOffentligTjenestepensjonResponse;
import no.nav.tjenestepensjon.simulering.Tjenestepensjonsimulering;
import no.nav.tjenestepensjon.simulering.consumer.TokenClient;
import no.nav.tjenestepensjon.simulering.domain.Stillingsprosent;
import no.nav.tjenestepensjon.simulering.domain.TPOrdning;
import no.nav.tjenestepensjon.simulering.rest.IncomingRequest;
import no.nav.tjenestepensjon.simulering.rest.OutgoingResponse.SimulertPensjon;

@Component
public class SoapClient extends WebServiceGatewaySupport implements Tjenestepensjonsimulering {

    private static final Logger LOG = LoggerFactory.getLogger(SoapClient.class);

    private final WebServiceTemplate webServiceTemplate;
    private final TokenClient tokenClient;

    public SoapClient(WebServiceTemplate template, TokenClient tokenClient) {
        this.webServiceTemplate = template;
        this.tokenClient = tokenClient;
    }

    @Override
    public List<Stillingsprosent> getStillingsprosenter(String fnr, TPOrdning tpOrdning) {
        var response = (HentStillingsprosentListeResponse) webServiceTemplate.marshalSendAndReceive(
                mapStillingsprosentRequest(fnr, tpOrdning),
                new StillingsprosentCallback(
                        "http://nav.no/ekstern/pensjon/tjenester/tjenestepensjonSimulering/v1/TjenestepensjonSimulering/hentStillingsprosentListeRequest",
                        tpOrdning.getTpLeverandor().getUrl(),
                        tokenClient.getSamlAccessToken().getAccessToken()));
        return mapStillingsprosentResponse(response);
    }

    @Override
    public List<SimulertPensjon> simulerPensjon(IncomingRequest incomingRequest, TPOrdning tpOrdning) {
        var response = (SimulerOffentligTjenestepensjonResponse) webServiceTemplate.marshalSendAndReceive(
                mapSimulerTjenestepensjonRequest(incomingRequest, tpOrdning),
                new StillingsprosentCallback(
                        "http://nav.no/ekstern/pensjon/tjenester/tjenestepensjonSimulering/v1/TjenestepensjonSimulering/simulerOffentligTjenestepensjonRequest",
                        tpOrdning.getTpLeverandor().getUrl(),
                        tokenClient.getSamlAccessToken().getAccessToken()));

        return mapSimulerTjenestepensjonResponse(response);
    }
}

File: src/main/kotlin/no/nav/tjenestepensjon/simulering/soap/SoapClient.kt
package no.nav.tjenestepensjon.simulering.soap

import no.nav.tjenestepensjon.simulering.Tjenestepensjonsimulering
import no.nav.tjenestepensjon.simulering.consumer.TokenClient
import no.nav.tjenestepensjon.simulering.domain.TpLeverandor
import no.nav.tjenestepensjon.simulering.model.v1.domain.FNR
import no.nav.tjenestepensjon.simulering.model.v1.domain.TPOrdning
import no.nav.tjenestepensjon.simulering.model.v1.request.HentStillingsprosentListeRequest
import no.nav.tjenestepensjon.simulering.model.v1.request.SimulerOffentligTjenestepensjonRequest
import no.nav.tjenestepensjon.simulering.model.v1.request.SimulerPensjonRequest
import no.nav.tjenestepensjon.simulering.model.v1.response.HentStillingsprosentListeResponse
import no.nav.tjenestepensjon.simulering.model.v1.response.SimulerOffentligTjenestepensjonResponse
import no.nav.tjenestepensjon.simulering.util.TPOrdningStillingsprosentMap
import org.slf4j.LoggerFactory
import org.springframework.beans.factory.annotation.Autowired
import org.springframework.beans.factory.annotation.Value
import org.springframework.stereotype.Component
import org.springframework.stereotype.Controller
import org.springframework.stereotype.Service
import org.springframework.ws.client.core.WebServiceTemplate
import org.springframework.ws.client.core.support.WebServiceGatewaySupport

@Service
@Controller
class SoapClient(
        webServiceTemplate: WebServiceTemplate,
        private val tokenClient: TokenClient
) : WebServiceGatewaySupport(), Tjenestepensjonsimulering {

    init {
        this.webServiceTemplate = webServiceTemplate
    }

    @Value("\${TJENESTEPENSJON_URL}")
    lateinit var simulerOffentlingTjenestepensjonUrl: String

    @Value("\${STILLINGSPROSENT_URL}")
    lateinit var hentStillingsprosentUrl: String

    @Value("\${SECURITY_CONTEXT_URL}")
    lateinit var samlSecurityContextUrl: String

    @Autowired
    lateinit var samlConfig: SamlConfig

    override fun getStillingsprosenter(
            fnr: FNR,
            tpOrdning: TPOrdning,
            tpLeverandor: TpLeverandor
    ) =
            (webServiceTemplate.marshalSendAndReceive(
                    HentStillingsprosentListeRequest(fnr, tpOrdning),
                    SOAPCallback(
                            hentStillingsprosentUrl,
                            tpLeverandor.url,
                            tokenClient.samlAccessToken.accessToken,
                            samlConfig
                    )
            ) as HentStillingsprosentListeResponse).stillingsprosentListe

    override fun simulerPensjon(
            request: SimulerPensjonRequest,
            tpOrdning: TPOrdning,
            tpLeverandor: TpLeverandor,
            tpOrdningStillingsprosentMap: TPOrdningStillingsprosentMap
    ) =
            (webServiceTemplate.marshalSendAndReceive(
                    with(request.simuleringsperioder) {
                        if (size > 1 && min()!!.isGradert())
                            SimulerOffentligTjenestepensjonRequest(
                                    simulerPensjonRequest = request,
                                    tpOrdning = tpOrdning,
                                    tpOrdningStillingsprosentMap = tpOrdningStillingsprosentMap,
                                    forsteUttak = min()!!,
                                    heltUttak = max()!!
                            )
                        else
                            SimulerOffentligTjenestepensjonRequest(
                                    simulerPensjonRequest = request,
                                    tpOrdning = tpOrdning,
                                    tpOrdningStillingsprosentMap = tpOrdningStillingsprosentMap,
                                    forsteUttak = min()!!
                            )
                    }.also { LOG.info("Mapped SimulerPensjonRequest: {} to SimulerOffentligTjenestepensjon: {}", request, it) },
                    SOAPCallback(
                            simulerOffentlingTjenestepensjonUrl,
                            tpLeverandor.url,
                            tokenClient.samlAccessToken.accessToken!!,
                            samlConfig
                    )
            ) as SimulerOffentligTjenestepensjonResponse).simulertPensjonListe

    companion object {
        @JvmStatic
        @Suppress("JAVA_CLASS_ON_COMPANION")
        private val LOG = LoggerFactory.getLogger(javaClass.enclosingClass)
    }

}